src/ccs.rs

```diff
@@ -0,0 +1,453 @@
+//! This module defines CCS related types and functions.
+#![allow(unused_imports)]
+#![allow(dead_code)]
+#![allow(clippy::type_complexity)]
+
+use crate::{
+  constants::{BN_LIMB_WIDTH, BN_N_LIMBS, NUM_FE_FOR_RO, NUM_HASH_BITS},
+  errors::NovaError,
+  gadgets::{
+    nonnative::{bignat::nat_to_limbs, util::f_to_nat},
+    utils::scalar_as_base,
+  },
+  r1cs::{R1CSInstance, R1CSShape, R1CSWitness, R1CS},
+  traits::{
+    commitment::CommitmentEngineTrait, commitment::CommitmentTrait, AbsorbInROTrait, Group, ROTrait,
+  },
+  utils::*,
+  Commitment, CommitmentKey, CE,
+};
+use bitvec::vec;
+use core::{cmp::max, marker::PhantomData};
+use ff::Field;
+use flate2::{write::ZlibEncoder, Compression};
+use itertools::concat;
+use rayon::prelude::*;
+use serde::{Deserialize, Serialize};
+use sha3::{Digest, Sha3_256};
+
+// TODO: Create a SparseMatrix type? Vec<(usize, usize, G::Scalar)>
+// TODO: Committed CCS using MLE (see src/spartan/pp.rs)
+// TODO: Linearized CCS struct and methods, separate struct similar to RelaxedR1CS
+
+/// Public parameters for a given CCS
+#[derive(Clone, Serialize, Deserialize)]
+#[serde(bound = "")]
+pub struct CCS<G: Group> {
+  _p: PhantomData<G>,
+}
+
+// NOTE: Currently m, n are implicit, could possibly infer from M
+/// A type that holds the shape of a CCS instance
+/// Unlike R1CS we have a list of matrices M instead of only A, B, C
+/// We also have t, q, d constants and c (vector), S (set)
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub struct CCSShape<G: Group> {
+  pub(crate) num_cons: usize,
+  pub(crate) num_vars: usize,
+  pub(crate) num_io: usize,
+  pub(crate) M: Vec<Vec<(usize, usize, G::Scalar)>>,
+  pub(crate) t: usize,
+  pub(crate) q: usize,
+  pub(crate) d: usize,
+  pub(crate) S: Vec<Vec<usize>>,
+  pub(crate) c: Vec<usize>,
+}
+
+/// A type that holds a witness for a given CCS instance
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub struct CCSWitness<G: Group> {
+  W: Vec<G::Scalar>,
+}
+
+// TODO: Make sure this is in the right form for committed CCS using MLE, possibly a separate type?
+/// A type that holds an CCS instance
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(bound = "")]
+pub struct CCSInstance<G: Group> {
+  pub(crate) comm_W: Commitment<G>,
+  pub(crate) X: Vec<G::Scalar>,
+}
+
+impl<G: Group> CCS<G> {
+  // TODO: Update commitment_key variables here? This is currently based on R1CS with M length
```

Comment:
This looks good! Sampling parameters to support the total number of non-zero entries should suffice for most finalizing/compressed snark.